fabtests/common/shared.c

```diff
@@ -3556,3 +3571,85 @@ void ft_free_string_array(char **s)
 	/* and then the actual array of pointers */
 	free(s);
 }
+
+static const char *nexttoken(const char *str,  int chr)
+{
+	if (str)
+		str = strchr(str, chr);
+	if (str)
+		str++;
+	return str;
+}
+
+static int ft_pin_core(const char *core_list)
+{
+	cpu_set_t mask;
+	size_t max_bits;
+	const char *curr_ptr, *next_ptr;
+	int r = 0;
+
+	max_bits = 8 * sizeof(mask);
+	next_ptr = core_list;
+	CPU_ZERO(&mask);
+
+	// parse each sub-list, delimited by comma
+	while (curr_ptr = next_ptr, next_ptr = nexttoken(next_ptr, ','), curr_ptr) {
+		int start, end;
+		const char *p1, *p2;
+		char c;
+		//get starting cpu number
+		if ((r = sscanf(curr_ptr, "%u%c", &start, &c)) < 1)
+			return EXIT_FAILURE;
+
+		end = start;
+		p1 = nexttoken(curr_ptr, '-');
+		p2 = nexttoken(curr_ptr, ',');
+		// get ending cpu number
+		if (p1 != NULL && (p2 == NULL || p1 < p2)) {
+			if ((r = sscanf(p1, "%u%c", &end, &c)) < 1)
+				return EXIT_FAILURE;
+		}
+
+		if (start > end)
+			return EXIT_FAILURE;
+		while (start <= end) {
+			if (start > max_bits)
+				return EXIT_FAILURE;
+			CPU_SET(start, &mask);
+			start++;
+		}
+	}
+
+	if (r == 2)
+		return EXIT_FAILURE;
+
+	return sched_setaffinity(0, sizeof(mask), &mask);
+}
+
+static int ft_parse_pin_core_opt(const char *opt, const char *valid_opt, char *optarg)
+{
+	char valid_opt_dash[32];
+
+	if (snprintf(valid_opt_dash, 32, "--%s", valid_opt) <= 0)
+		return EXIT_FAILURE;
+
+	if (strcmp(opt, valid_opt_dash) != 0) {
+		printf("Option invalid, use %s <core number>\n", valid_opt_dash);
```

Comment:
The print shows that the option takes a single input value, core_number, but the code does some complex string parsing.